autotune: enable autotune in mission mode

mahima-yoga opened the pull request:
### Solved Problem

When operating fully autonomously (i.e., without manual control or a guaranteed link to the ground station), tuning a vehicle can be difficult since the current autotune process requires either RC input or a GCS command. For these scenarios, it would be useful to trigger autotune as part of a mission.

Current behavior:

- Multicopter (MC): Autotune is enabled via a GCS button, which sends the MAVLink command `MAV_CMD_DO_AUTOTUNE_ENABLE`. This command is handled by the `mavlink_receiver`, which sets the parameter `MC_AT_START = true`. The MC autotune module detects this parameter change and starts the FSM.

- Fixed-wing (FW): Autotune can be enabled via an RC AUX switch or by manually setting the parameter `FW_AT_START = true`. The FW autotune module monitors this parameter and the aux switch and runs the FSM when either are enabled.

- Limitation: Neither implementation supports triggering autotune as part of a mission, and the `*_AT_START` parameter handling can be cleaned up. 

### Solution

- Extend mission support: Add support for the `MAV_CMD_DO_AUTOTUNE_ENABLE` command within the mission framework, allowing autotune to be triggered as a mission item.

- Simplify logic: Remove the MC_AT_START and FW_AT_START parameters.

- Unified command handling: Add `CMD_DO_AUTOTUNE_ENABLE` to `VehicleCommand.msg` and have the autotune modules subscribe directly to this message.

This ensures that autotune can be triggered consistently: whether from a mission or a GCS command.  

### Changelog Entry
For release notes:
```
Feature/Bugfix Support running Autotune as a mission item
```

### Test coverage

- [x] Verified MC autotune can be triggered from mission item in SITL
- [x] Verified FW autotune can be triggered from mission item in SITL 
- [x] Confirmed MC GCS-triggered autotune still works in SITL 
- [x] Confirmed FW AUX-triggered autotune still works in SITL 
- [x] Flight-tested 

### Context
Related links, screenshot before/after, video

hamishwillee commented:
@mahima-yoga 
1. FMI, why would you want to?
2. I recently created https://github.com/PX4/PX4-Autopilot/pull/25795 to fix up the command handling. The same changes would need to be considered here (and may be already). Specifically
   - The first parameter is enable (1) or disable (0). The code should obey both those values (as a command, it currently ignores param1 and enables, hence my post). The code should reject an item with any other value for param1
   - The second param should allow you to set the axis for tuning. Currently that is ignored. Either this must be 0 and any other value should result in a NACK OR we should support the axis setting (on FW). I'd like to do that, but I THINK to make that happen the values have to be transient in a mission or until mode change/reboot.

IF you want to include my fixups in your PR that would be great and I could close it. Not a programmer - just wanted to MAVLink command to work as designed.

sfuhrer reviewed (commented):
Nice cleanup and minimally invasive way to add a big new feature (auto-tuning in mission mode, fully autonomous). 

To address @hamishwillee 's valid [concerns](https://github.com/PX4/PX4-Autopilot/pull/25797#issuecomment-3438681163):
I would make sure that autotuning is aborted when the mode is switched out of mission. But not spend too much effort in trying to be smart and only allow auto tuning in certain flight conditions. Rather that should be the operators responsibility - for that docs extensions are important.

hamishwillee commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -622,42 +622,33 @@ void MavlinkReceiver::handle_message_command_both(mavlink_message_t *msg, const
 		autotune_attitude_control_status_s status{};
```
I think we should check command validity here - i.e. if we're touching this anyway, pull the changes I put in around https://github.com/PX4/PX4-Autopilot/pull/25795/files#r2449948304 that check the result and cause the ACK ( but with your much better value checking)

See also https://github.com/PX4/PX4-Autopilot/pull/25797#discussion_r2476111958

hamishwillee commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -622,42 +622,33 @@ void MavlinkReceiver::handle_message_command_both(mavlink_message_t *msg, const
 				case vehicle_status_s::VEHICLE_TYPE_ROTARY_WING:
-					atune_start = param_find("MC_AT_START");
-
+					has_module = param_find("MC_AT_APPLY");
```
Thanks - yes we don't need to address "1". We still want to document what is supported though.

Such as add the command to lists in https://docs.px4.io/main/en/flight_modes_mc/mission#mission_commands and https://docs.px4.io/main/en/flight_modes_fw/mission#mission-commands 

In the autotune docs, see if anything needs to be updated after this:
- https://docs.px4.io/main/en/config/autotune_mc
- https://docs.px4.io/main/en/config/autotune_fw

mahima-yoga replied:
Thanks! I have updated the mission docs to include the autotune command and it's "caveats". Not much has changed in terms of user experience in the actual autotune module, so I don't think we require a documentation update there

hamishwillee replied:
The docs changes seem good to me thanks!

hamishwillee commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -622,46 +622,41 @@ void MavlinkReceiver::handle_message_command_both(mavlink_message_t *msg, const
+
+			if (fabsf(vehicle_command.param2) > FLT_EPSILON) {
+				PX4_WARN("Autotune axis selection not supported through mavlink. Use FW_AT_AXES to set axes for fixed-wing vehicles.");
+			}
```
@mahima-yoga Me reviewing code is very unreliable. My understanding here though is that
- if the module exists we publish the command with the values from the mavlink message for param1 and param2.
  That should start the autotune operation, which is executed in the autotune code.
- If param1 is exactly 1 (autotune) we do the mavlink progress notification - i.e. we know that we started autotuning.
  - We also warn here if param2 is a value other than 0.
     > [!NOTE]
     > PX4 should NACK the request with [MAV_RESULT_DENIED](https://mavlink.io/en/messages/common.html#MAV_RESULT_DENIED) because the parameters are wrong for this case. Why? Because users won't know that the system is ignoring the value.
- If module is not installed or param1 is other than 1 we NACK with VEHICLE_CMD_RESULT_UNSUPPORTED.
  That's incorrect. We should NACK with VEHICLE_CMD_RESULT_UNSUPPORTED if the module is not installed, and `MAV_RESULT_DENIED` if the value is other than 1.
 
 
 Upshot, if any parameter is wrong, then it should be MAV_RESULT_DENIED, but if PX4 can't execute because it doesn't understand the command then it should NACK VEHICLE_CMD_RESULT_UNSUPPORTED

hamishwillee replied:
@mahima-yoga So close!

- has module but incorrect param1 **or param2 anything other than 0**: denied NACK

This ensures that broken missions where the value is set to a value we don't support don't succeed.

mahima-yoga replied:
@hamishwillee does this look fine to you now? Would like to merge this asap

hamishwillee replied:
With the caveat that I am not a great reviewer, this looks perfect to me. Thanks very much.

mahima-yoga commented on src/modules/fw_autotune_attitude_control/fw_autotune_attitude_control.cpp:
```diff
@@ -107,10 +101,23 @@ void FwAutotuneAttitudeControl::Run()
+				}
+			}
+		}
+	}
```
Sorry yes - mavlink receiver, typo

Currently untested, and would still require the small addition inside the autotune modules themselves to call the stopAutotune() functions respectively. If it's ok with you I would like to keep this PR with the current scope, and then we can open a new one for disabling over mavlink.

sfuhrer reviewed (dismissed):
My comments were addressed. Please squash merge.

hamishwillee commented on docs/en/flight_modes_fw/mission.md:
```diff
@@ -162,6 +162,9 @@ Mission Items:
 - [MAV_CMD_DO_SET_CAMERA_FOCUS](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_SET_CAMERA_FOCUS)
+- [MAV_CMD_DO_AUTOTUNE_ENABLE](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_AUTOTUNE_ENABLE)
+   - Disabling autotune by setting `param1` to zero is currently not supported. To abort autotune during a mission, switch to another flight mode.
+   - Axis selection specified in the MAVLink message (`param2`) is ignored. Instead, the axis bitmask defined by [`FW_AT_AXES`](../advanced_config/parameter_reference.md#FW_AT_AXES) is used.
```
```suggestion
   - Axis selection specified in the MAVLink message is ignored (`param2` must be set to 0).
     Instead, the axis bitmask defined by [`FW_AT_AXES`](../advanced_config/parameter_reference.md#FW_AT_AXES) is used.
```

hamishwillee reviewed (dismissed):
LGTM. Don't forget to squash-merge. 

Right now I am ~expecting~ hoping for a follow on PR for command that stops tune as well. Let me know if you decide _not_ to do it.

sfuhrer reviewed (approved):
LGTM

mahima-yoga committed:
autotune: enable autotune in mission mode

When operating fully autonomously (i.e., without manual control or a guaranteed link to the ground station), tuning a vehicle can be difficult since the current autotune process requires either RC input or a GCS command. For these scenarios, it it useful to be able to start autotune inside a mission.

mahima-yoga committed:
v6s: remove rover to free up flash
